patches/0023-Add-support-for-respecting-system-proxy-settings-whe.patch

```diff
@@ -0,0 +1,504 @@
++#elif defined(OS_LINUX)
++    return std::make_unique<ProxyResolverFactoryLinux>();
+ #else
+     NOTREACHED();
+     return nullptr;
+@@ -265,7 +270,7 @@ class ProxyResolverFactoryForSystem : public MultiThreadedProxyResolverFactory {
+   }
+ 
+   static bool IsSupported() {
+-#if defined(OS_WIN) || defined(OS_APPLE)
++#if defined(OS_WIN) || defined(OS_APPLE) || defined(OS_LINUX)
+     return true;
+ #else
+     return false;
+diff --git a/net/proxy_resolution/proxy_config_service_linux.cc b/net/proxy_resolution/proxy_config_service_linux.cc
+index c11685ff4f221..cba0225d69512 100644
+--- a/net/proxy_resolution/proxy_config_service_linux.cc
++++ b/net/proxy_resolution/proxy_config_service_linux.cc
+@@ -31,6 +31,7 @@
+ #include "base/threading/thread_restrictions.h"
+ #include "base/timer/timer.h"
+ #include "net/base/proxy_server.h"
++#include "sandbox/linux/services/flatpak_sandbox.h"
+ 
+ #if defined(USE_GIO)
+ #include <gio/gio.h>
+@@ -1042,11 +1043,18 @@ ProxyConfigServiceLinux::Delegate::GetConfigFromSettings() {
+   ProxyConfig config;
+ 
+   std::string mode;
+-  if (!setting_getter_->GetString(SettingGetter::PROXY_MODE, &mode)) {
++  bool use_flatpak_sandbox = (sandbox::FlatpakSandbox::GetInstance()->GetSandboxLevel() >
++                              sandbox::FlatpakSandbox::SandboxLevel::kNone);
++
++  if (!use_flatpak_sandbox &&
++      !setting_getter_->GetString(SettingGetter::PROXY_MODE, &mode)) {
+     // We expect this to always be set, so if we don't see it then we probably
+     // have a gsettings problem, and so we don't have a valid proxy config.
+     return base::nullopt;
++  } else if (use_flatpak_sandbox) {
++    mode = "auto";
+   }
++
+   if (mode == "none") {
+     // Specifically specifies no proxy.
+     return ProxyConfigWithAnnotation(
+diff --git a/net/proxy_resolution/proxy_resolver_linux.cc b/net/proxy_resolution/proxy_resolver_linux.cc
+new file mode 100644
+index 0000000000000..93e2f5c2b399e
+--- /dev/null
++++ b/net/proxy_resolution/proxy_resolver_linux.cc
+@@ -0,0 +1,234 @@
++// Copyright (c) 2020 Endless OS Foundation LLC
++// Use of this source code is governed by a BSD-style license that can be
++// found in the LICENSE file.
++
++#include "net/proxy_resolution/proxy_resolver_linux.h"
++
++#include "base/message_loop/message_pump_type.h"
++#include "base/no_destructor.h"
++#include "base/strings/string_util.h"
++#include "base/synchronization/waitable_event.h"
++#include "base/threading/thread_task_runner_handle.h"
++#include "base/threading/thread.h"
++#include "base/time/time.h"
++#include "base/timer/timer.h"
++#include "net/base/net_errors.h"
++#include "net/proxy_resolution/proxy_info.h"
++#include "net/proxy_resolution/proxy_resolver.h"
++#include "sandbox/linux/services/flatpak_sandbox.h"
++
++#if defined(USE_GIO)
++#include <gio/gio.h>
++#endif  // defined(USE_GIO)
++
++namespace net {
++
++class NetworkIsolationKey;
++
++namespace {
++
++class GLibThread {
```

Comment:
Question: do we actually need a separate thread here? GLib declares global structures to be thread-safe, so I believe `g_proxy_resolver_lookup_async` could be called from the same thread, locking the cache if necessary.

Reply:
Tbh I wasn't sure either and on a **quick** look at the code (portal impls) they didn't look thread safe to me so I went with the safest option. Maybe @pwithnall would be able to confirm/clarify here.

Reply:
@refi64 is right: you should be able to expect that the proxy resolver returned by `g_proxy_resolver_get_default()` is thread-safe.

The `gsimpleproxyresolver.c` implementation is thread-safe as long as there are no calls to `g_simple_proxy_resolver_set_*()` after any threads have been spawned.

The `gproxyresolverportal.c` implementation is entirely thread-safe, because it just passes calls through to a `GDBusProxy` around the portal, which is thread-safe.

The `gproxyresolvergnome.c` implementation (in glib-networking) looks like it’s thread-safe (it has a mutex), but I haven’t checked it thoroughly.

The `glibproxyresolver.c` (libproxy) implementation (in glib-networking) also looks like it has thought about threading, because it already does its libproxy calls in a worker thread.

tl;dr: You should be able to drop your worker thread here. If any threading bugs do appear, they are likely bugs in GLib; but I wouldn’t expect any to appear if you’re using any implementation except `gsimpleproxyresolver.c`. It would be unlikely you’d end up using that implementation, as it’s a fallback.

Reply:
Thanks @pwithnall, @refi64. I've updated the impl to drop the glib worker thread. Some possible improvements using current GProxyResolver/Portal interface APIs (for the future - I don't plan to work on them for the time being):
- avoid calls to `g_proxy_resolver_lookup()` if there is an existing inflight request for the same url
  - although `net::ProxyResolver::GetProxyForURL()` is always invoked from the same thread for the same `net::ProxyResolver` instance, different instances may be invoked from different threads but still use the same `GProxyResolver` instance...
- using a single/global cache for all proxy resolver instances
  - as above, given we use the same `GProxyResolver` instance behind the sceces, we could consider a global cache
- make cache smarter
- allow async `net::ProxyResolver::GetProxyForURL()` calls (although currently the only consumer is `net::MultiThreadProxyResolver` which always expects a sync call)